File: src/rmpp_enhancer/pipeline.py
"""
Core image processing pipeline for reMarkable Paper Pro.

Features:
- Exact 1:1 pixel mapping for 11.8" Gallery 3 Canvas Color screen (2160 x 1620 @ 229 PPI).
- Calibrated 3D LUT inverse compensation (OKLab v3 color space).
- Bilateral edge-directed inking filter to deepen dialogue line-art and counteract pigment dithering.
- Tuned JPEG compression for fast cloud sync and minimal storage.
"""

import os
import threading
from dataclasses import dataclass
from typing import Optional
from PIL import Image, ImageFilter
import numpy as np

from rmpp_enhancer.profiles import DEFAULT_LUT_PATH


@dataclass
class EnhancerConfig:
    quality: int = 82               # Tuned sweet-spot for fast reMarkable sync
    subsampling: int = 0            # 0 = 4:4:4 (full chroma), 2 = 4:2:0
    color_correction: bool = True   # Apply OKLab v3 3D LUT
    edge_inking: bool = True        # Apply bilateral edge-directed inking
    lut_path: Optional[str] = None  # Path to .cube LUT file (None = use bundled)
    target_width: int = 1620        # RMPP portrait width
    target_height: int = 2160       # RMPP portrait height
    dpi: int = 229                  # Native screen density


# Global cache for the parsed Pillow 3D LUT. Pages are processed on a thread
# pool, so the cache is guarded to keep parsing to a single pass.
_CACHED_LUT: Optional[ImageFilter.Color3DLUT] = None
_CACHED_LUT_PATH: Optional[str] = None
_LUT_LOCK = threading.Lock()


def load_3d_lut(cube_path: Optional[str] = None) -> ImageFilter.Color3DLUT:
    """Loads and caches a .cube 3D LUT into Pillow's Color3DLUT format."""
    actual_path = cube_path or DEFAULT_LUT_PATH

    if _CACHED_LUT is not None and _CACHED_LUT_PATH == actual_path:
        return _CACHED_LUT

    with _LUT_LOCK:
        return _parse_and_cache_lut(actual_path)


def _parse_and_cache_lut(actual_path: str) -> ImageFilter.Color3DLUT:
    global _CACHED_LUT, _CACHED_LUT_PATH

    if _CACHED_LUT is not None and _CACHED_LUT_PATH == actual_path:
        return _CACHED_LUT

    if not os.path.exists(actual_path):
        raise FileNotFoundError(f"3D LUT profile not found at {actual_path}")

    lut_table = []
    size = 33
    with open(actual_path, "r", encoding="utf-8", errors="replace") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if line.startswith("LUT_3D_SIZE"):
                size = int(line.split()[1])
                continue
            if line.startswith("TITLE") or line.startswith("DOMAIN"):
                continue
            parts = [float(x) for x in line.split()]
            if len(parts) == 3:
                lut_table.extend(parts)

    expected_len = size * size * size * 3
    if len(lut_table) != expected_len:
        raise ValueError(f"Invalid LUT data in {actual_path}: expected {expected_len} values, got {len(lut_table)}")

    _CACHED_LUT = ImageFilter.Color3DLUT(size, lut_table)
    _CACHED_LUT_PATH = actual_path
    return _CACHED_LUT


def prepare_rgb(img: Image.Image) -> Image.Image:
    """Ensures image is in RGB mode with alpha composited over pure white."""
    if img.mode == "RGB":
        return img
    if img.mode in ("RGBA", "LA") or (img.mode == "P" and "transparency" in img.info):
        bg = Image.new("RGB", img.size, (255, 255, 255))
        rgba = img.convert("RGBA")
        bg.paste(rgba, mask=rgba.split()[3])
        return bg
    return img.convert("RGB")


def scale_to_rmpp_geometry(img: Image.Image, config: EnhancerConfig) -> Image.Image:
    """Scales image proportionally to reMarkable Paper Pro Option A screen dimensions."""
    w, h = img.size
    if h >= w:
        # Portrait: fit within 1620 x 2160 (height target 2160)
        scale = min(config.target_width / w, config.target_height / h)
    else:
        # Landscape spread: fit within 2160 x 1620
        scale = min(config.target_height / w, config.target_width / h)

    new_w = max(1, int(round(w * scale)))
    new_h = max(1, int(round(h * scale)))

    if (new_w, new_h) != (w, h):
        return img.resize((new_w, new_h), Image.Resampling.LANCZOS)
    return img


_LUMA_WEIGHTS = np.array([0.299, 0.587, 0.114], dtype=np.float32)


def apply_edge_directed_inking(img: Image.Image) -> Image.Image:
    """
    Applies bilateral edge-directed inking filter:
    1. Detects high-contrast edge gradients.
    2. Deepens dark ink lines (lum < 0.40) to counteract pigment diffusion.
    3. Brightens edge halos (lum >= 0.40) for crisp line separation.
    4. Applies micro-contrast unsharp mask.
    """
    gray = img.convert("L")
    edges = gray.filter(ImageFilter.FIND_EDGES)
    edges_arr = np.asarray(edges, dtype=np.float32) / 255.0
    edge_mask = edges_arr > 0.16

    arr = np.asarray(img, dtype=np.float32) / 255.0
    lum = arr @ _LUMA_WEIGHTS

    # Single per-pixel gain map: 0.75 on dark ink, 1.10 on the lighter side of an
    # edge, 1.0 everywhere else. Applied in one pass over all three channels.
    gain = np.where(edge_mask, np.where(lum < 0.40, 0.75, 1.10), 1.0).astype(np.float32)
    arr *= gain[..., None]
    np.clip(arr, 0.0, 1.0, out=arr)

    res = Image.fromarray(np.rint(arr * 255).astype(np.uint8))
    return res.filter(ImageFilter.UnsharpMask(radius=1.0, percent=115, threshold=3))


def process_image(img: Image.Image, config: EnhancerConfig) -> Image.Image:
    """Runs a single page image through the full RMPP processing pipeline."""
    rgb = prepare_rgb(img)

    # 1. Scale to Option A Geometry (@ 229 PPI)
    scaled = scale_to_rmpp_geometry(rgb, config)

    # 2. 3D LUT Color Calibration
    pil_lut = load_3d_lut(config.lut_path) if config.color_correction else None
    if config.color_correction and pil_lut is not None:
        corrected = scaled.filter(pil_lut)
    else:
        corrected = scaled

    # 3. Bilateral Edge Inking Filter
    if config.edge_inking:
        final_img = apply_edge_directed_inking(corrected)
    else:
        final_img = corrected

    return final_img


def save_page_jpeg(img: Image.Image, dst_path: str, config: EnhancerConfig) -> str:
    """Saves enhanced image to JPEG with specified quality, chroma subsampling, and DPI."""
    img.save(
        dst_path,
        "JPEG",
        quality=config.quality,
        subsampling=config.subsampling,
        optimize=True,
        dpi=(config.dpi, config.dpi),
    )
    return dst_path

File: rust/src/pipeline.rs
//! Core image processing pipeline for reMarkable Paper Pro.

use image::{DynamicImage, Rgb, RgbImage};

use fast_image_resize as fir;
use jpeg_encoder::{ColorType, Encoder, PixelDensity, SamplingFactor};

use crate::inking::apply_edge_directed_inking;
use crate::lut::Lut3D;

#[derive(Clone, Debug)]
pub struct EnhancerConfig {
    pub quality: u8,            // Sweet-spot 82 for cloud sync
    pub subsampling: u8,        // 0 = 4:4:4, 2 = 4:2:0
    pub color_correction: bool, // OKLab v3 3D LUT
    pub edge_inking: bool,      // Bilateral edge-directed inking
    pub lut_path: Option<String>,
    pub target_width: u32,  // 1620
    pub target_height: u32, // 2160
    pub dpi: u16,           // 229
}

impl Default for EnhancerConfig {
    fn default() -> Self {
        Self {
            quality: 82,
            subsampling: 0,
            color_correction: true,
            edge_inking: true,
            lut_path: None,
            target_width: 1620,
            target_height: 2160,
            dpi: 229,
        }
    }
}

/// Ensures image is in RGB mode with alpha composited over pure white.
pub fn prepare_rgb(img: DynamicImage) -> RgbImage {
    match img {
        DynamicImage::ImageRgb8(rgb) => rgb,
        other => {
            let rgba = other.to_rgba8();
            let (w, h) = rgba.dimensions();
            let mut out = RgbImage::new(w, h);
            for (x, y, pixel) in rgba.enumerate_pixels() {
                let a = pixel[3] as f32 / 255.0;
                if a >= 1.0 {
                    out.put_pixel(x, y, Rgb([pixel[0], pixel[1], pixel[2]]));
                } else if a <= 0.0 {
                    out.put_pixel(x, y, Rgb([255, 255, 255]));
                } else {
                    let r = ((pixel[0] as f32 * a) + 255.0 * (1.0 - a)).round() as u8;
                    let g = ((pixel[1] as f32 * a) + 255.0 * (1.0 - a)).round() as u8;
                    let b = ((pixel[2] as f32 * a) + 255.0 * (1.0 - a)).round() as u8;
                    out.put_pixel(x, y, Rgb([r, g, b]));
                }
            }
            out
        }
    }
}

/// Computes Option A target dimensions for portrait or landscape orientation.
pub fn compute_target_geometry(w: u32, h: u32, config: &EnhancerConfig) -> (u32, u32) {
    let scale = if h >= w {
        // Portrait: fit within 1620 x 2160 (height target 2160)
        (config.target_width as f64 / w as f64).min(config.target_height as f64 / h as f64)
    } else {
        // Landscape spread: fit within 2160 x 1620
        (config.target_height as f64 / w as f64).min(config.target_width as f64 / h as f64)
    };

    let new_w = (w as f64 * scale).round().max(1.0) as u32;
    let new_h = (h as f64 * scale).round().max(1.0) as u32;
    (new_w, new_h)
}

/// Scales image proportionally to Option A screen geometry using SIMD Lanczos3.
pub fn scale_to_rmpp_geometry(img: &RgbImage, config: &EnhancerConfig) -> RgbImage {
    let (w, h) = img.dimensions();
    let (new_w, new_h) = compute_target_geometry(w, h, config);

    if (new_w, new_h) == (w, h) {
        return img.clone();
    }

    let src_view = fir::images::ImageRef::new(w, h, img.as_raw(), fir::PixelType::U8x3)
        .expect("Invalid source image buffer");
    let mut dst_image = fir::images::Image::new(new_w, new_h, fir::PixelType::U8x3);
    let mut resizer = fir::Resizer::new();
    let options = fir::ResizeOptions::new()
        .resize_alg(fir::ResizeAlg::Interpolation(fir::FilterType::Lanczos3));

    resizer
        .resize(&src_view, &mut dst_image, &options)
        .expect("Failed to resize image");

    RgbImage::from_raw(new_w, new_h, dst_image.into_vec())
        .expect("Failed to construct RgbImage from resized buffer")
}

/// Runs a single page through scaling, 3D LUT calibration, and edge-directed inking.
pub fn process_image(img: DynamicImage, config: &EnhancerConfig, lut: Option<&Lut3D>) -> RgbImage {
    let rgb = prepare_rgb(img);

    // 1. Scale to Option A Geometry (@ 229 PPI)
    let mut scaled = scale_to_rmpp_geometry(&rgb, config);

    // 2. 3D LUT Color Calibration
    if config.color_correction {
        if let Some(l) = lut {
            l.apply(&mut scaled);
        }
    }

    // 3. Bilateral Edge Inking Filter
    if config.edge_inking {
        apply_edge_directed_inking(&scaled)
    } else {
        scaled
    }
}

/// Encodes an RgbImage directly into JPEG bytes with tuned quality, chroma subsampling, and DPI.
pub fn encode_page_jpeg(img: &RgbImage, config: &EnhancerConfig) -> Result<Vec<u8>, String> {
    let mut buf = Vec::new();
    let mut encoder = Encoder::new(&mut buf, config.quality);
    let sampling = match config.subsampling {
        0 => SamplingFactor::R_4_4_4,
        2 => SamplingFactor::R_4_2_0,
        _ => SamplingFactor::R_4_4_4,
    };
    encoder.set_sampling_factor(sampling);
    encoder.set_density(PixelDensity::dpi(config.dpi));
    let (w, h) = img.dimensions();
    encoder
        .encode(img.as_raw(), w as u16, h as u16, ColorType::Rgb)
        .map_err(|e| format!("JPEG encoding error: {}", e))?;
    Ok(buf)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_option_a_scaling_portrait() {
        let config = EnhancerConfig::default();
        let (w, h) = compute_target_geometry(1080, 1920, &config);
        assert_eq!(h, 2160);
        assert_eq!(w, 1215);

        let img = RgbImage::new(1080, 1920);
        let scaled = scale_to_rmpp_geometry(&img, &config);
        assert_eq!(scaled.dimensions(), (1215, 2160));
    }

    #[test]
    fn test_option_a_scaling_landscape() {
        let config = EnhancerConfig::default();
        let (w, h) = compute_target_geometry(1920, 1080, &config);
        assert!(w <= 2160 && h <= 1620);
        assert_eq!(w, 2160);
        assert_eq!(h, 1215);

        let img = RgbImage::new(1920, 1080);
        let scaled = scale_to_rmpp_geometry(&img, &config);
        assert_eq!(scaled.dimensions(), (2160, 1215));
    }

    #[test]
    fn test_prepare_rgb_alpha() {
        let mut rgba = image::RgbaImage::new(10, 10);
        for pixel in rgba.pixels_mut() {
            *pixel = image::Rgba([255, 0, 0, 0]); // fully transparent
        }
        let rgb = prepare_rgb(DynamicImage::ImageRgba8(rgba));
        assert_eq!(rgb.get_pixel(5, 5), &Rgb([255, 255, 255]));
    }
}
